### PR TITLE
[Feat] Improve browser notifications

### PR DESCRIPTION
**Pull request by antonyayoub**
# Name of the feature/issue
Improve browser notifications
## Description
This PR improves browser notifications by specifying the transaction category (Swap, Bridge, Approval & so on ....) and details (amount, network name & so on..) in each notification with its status (completed, cancelled, failed).

### CODE REVIEW

File: packages/background/src/utils/notifications.ts [hunk 1]
```diff
@@ -1,217 +1,267 @@
...
-const linkToEtherscan = (url: string) => {
-    if (url.startsWith('https://')) {
-        chrome.tabs.create({ url: url });
+    const isTxFailed = status === TransactionStatus.FAILED;
+    const isTxCancelled = status === TransactionStatus.CANCELLED;
+
+    // Used when the transaction category is not supported or there is no data for the transaction category.
+    let showDefaultNotification = false;
+
+    switch (transactionCategory) {
```
Comment (julianariel) [inline packages/background/src/utils/notifications.ts:140]: the only improvement I can imagine here is having a collection of notification content by type and accessing it using the category as an index, then replacing the parameters with actual data. Typing could be something like this : 
```typescript
type Notification = {
    title: string,
    message: string,
    url: string
}

type NotificationsContent = {
    [key: string]: Notification;
};


const NotificationContent: NotificationsContent = {
    'exchange': {
        title: "Swap",
        message: "${fromToken} to ${toToken} swap on ${network}",
        url: ""
    },
}    
``` 

but even without this, logic is quite clear and concise 👏🏻

**Review by julianariel — APPROVED** [review]
Code looks great, good work 🔥 added small suggestions only. 

I'll test the functionality next!

File: packages/background/src/controllers/transactions/TransactionController.ts [hunk 2]
```diff
@@ -428,8 +428,11 @@ export class TransactionController extends BaseController<
             TransactionEvents.STATUS_UPDATE,
             (transactionMeta: TransactionMeta) => {
                 if (
-                    transactionMeta.status === TransactionStatus.CONFIRMED ||
-                    transactionMeta.status === TransactionStatus.FAILED
+                    this._preferencesController.settings
+                        .subscribedToNotifications &&
+                    (transactionMeta.status === TransactionStatus.CONFIRMED ||
```
[the remark this answers is not in the document]
Comment (antonyayoub) [reply packages/background/src/controllers/transactions/TransactionController.ts:433]: Great catch 👌

File: packages/background/src/utils/token.ts [hunk 3]
```diff
@@ -12,6 +12,7 @@ import {
 } from './types/ethereum';
 import { BigNumber } from 'ethers';
 import { MaxUint256 } from '@ethersproject/constants';
+import { formatUnits } from 'ethers/lib/utils';
```
[the remark this answers is not in the document]
Comment (antonyayoub) [reply packages/background/src/utils/token.ts:15]: Always forget to use that 😓

File: packages/background/src/utils/notifications.ts [hunk 4]
```diff
@@ -1,217 +1,267 @@
...
+/**
+ * Shows a browser notification and open link on click.
+ *
+ * @param title - The notification title.
+ * @param message - The notification message.
+ * @param url - The url to open on click.
+ *
+ */
+const showNotification = (title: string, message: string, url: string) => {
+    if (url) addOnClickListener();
```
[the remark this answers is not in the document]
Comment (antonyayoub) [reply packages/background/src/utils/notifications.ts:58]: It actually does work as it adds a dynamic function (listener) called linkToExplorer to the notifications. The function takes the notificationID (which is the URL in our case and then uses it to create the new tab only if the url param starts with "https://").

File: packages/background/src/utils/notifications.ts [hunk 5]
```diff
@@ -1,217 +1,267 @@
...
+                value,
+                decimals,
+                symbol
+            )} on ${txNetworkName}.`;
+            break;
+        }
+        case TransactionCategories.CONTRACT_INTERACTION:
+            title = `Contract Interaction`;
+            message = `Interaction with contract on ${txNetworkName}`;
+            break;
```
[the remark this answers is not in the document]
Comment (antonyayoub) [reply packages/background/src/utils/notifications.ts:227]: You're right. It's currently just like the photo below.
Screenshot 2023-02-23 at 6 40 27 PM

What do you think about adding the origin or/and method signature to the message somewhere?

[the remark this answers is not in the document]
Comment (antonyayoub) [reply packages/background/src/utils/notifications.ts:227]: Nope, only the whole origin.

File: packages/background/src/utils/notifications.ts [hunk 6]
```diff
@@ -1,217 +1,267 @@
...
+ *
+ */
+const showNotification = (title: string, message: string, url: string) => {
+    if (url) addOnClickListener();
 
-    const url = createCustomExplorerLink(hash as string, explorerUrl);
-    const title = 'Transaction confirmed';
-    const message = `Transaction with nonce ${nonce} confirmed!`;
+    // To prevent duplicate notifications id which causes the notification to not show (overrides the old one)
+    const notificationUrl = url + `?timestamp=${Date.now()}`;
```
[the remark this answers is not in the document]
Comment (antonyayoub) [reply packages/background/src/utils/notifications.ts]: done 💪

File: packages/background/src/utils/token.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (antonyayoub) [reply packages/background/src/utils/token.ts]: done 💪

File: packages/background/src/controllers/transactions/TransactionController.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (antonyayoub) [reply packages/background/src/controllers/transactions/TransactionController.ts]: done 💪

File: packages/background/src/utils/notifications.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (julianariel) [reply packages/background/src/utils/notifications.ts:240]: we can try parsing the whole origin and extracting the domain name, or re-using the same spender list we consume for the token allowances feature that is basically the same contract interaction

File: packages/background/src/utils/notifications.ts [hunk 7]
```diff
@@ -1,217 +1,267 @@
...
-
-export const showIncomingTransactionNotification = (
-    account: string,
-    chainId: number,
-    section?: '' | 'tokentxns' | 'tokentxnsErc721' | 'tokentxnsErc1155'
-): void => {
-    const explorerUrl = getExplorerUrl(chainId);
-    if (!explorerUrl) {
+        txMeta.metaType === MetaType.CANCEL ||
+        txMeta.transactionCategory === TransactionCategories.INCOMING ||
```
[the remark this answers is not in the document]
Comment (antonyayoub) [reply packages/background/src/utils/notifications.ts]: That was an error now it's fixed great catch! 👌

File: packages/background/src/utils/notifications.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (antonyayoub) [reply packages/background/src/utils/notifications.ts:271]: We now add either the Contract name if exists in our DApp list or the contract address in this form "Contract (0x23fw...4gf2)".

**Pull request merged**